src/middleware/FileRestrictionsMiddleware.ts

```diff
@@ -0,0 +1,108 @@
+                .unsafeCoerce();
+
+        const mAppeal: Maybe<Appeal> = session.chain(_ => _.getExtraData())
+            .chainNullable<ApplicationData>(extraData => extraData[APPLICATION_DATA_KEY])
+            .chainNullable(appData => appData.appeal)
+            .ifNothing(() => loggerInstance().error(`${FileRestrictionsMiddleware.name} - Appeal was expected in session but none found`));
+
+        const fileId: string = req.params.fileId;
+
+        const hasAdminPermissions: boolean = this.hasAdminPermissions(signInInfo);
+        const hasUserPermissions: boolean = this.hasUserPermissions(fileId, signInInfo, mAppeal.unsafeCoerce());
+
+        return hasAdminPermissions || hasUserPermissions ? next() : this.renderForbiddenError(res);
+    }
+
+    private hasAdminPermissions(signInInfo: ISignInInfo): boolean {
+
+        const userProfile = signInInfo[SignInInfoKeys.UserProfile];
+
+        if (!userProfile) {
+            throw new Error(`${FileRestrictionsMiddleware.name} - User profile was expected in session but none found`);
+        }
+
+        const adminPermissionFlag: string | undefined = signInInfo[SignInInfoKeys.AdminPermissions];
+        const permissions = userProfile[UserProfileKeys.Permissions];
+
+        if (!adminPermissionFlag || adminPermissionFlag !== '1') {
+            return false;
+        }
+
+        return permissions &&
+            permissions[AppealsPermissionKeys.download] &&
+            permissions[AppealsPermissionKeys.view];
+    }
+
+    private hasUserPermissions(fileId: string, signInInfo: ISignInInfo, appeal: Appeal): boolean {
+
+        const userProfile = signInInfo[SignInInfoKeys.UserProfile];
+
+        if (!userProfile) {
+            throw new Error(`${FileRestrictionsMiddleware.name} - User profile was expected in session but none found`);
+        }
+
+        const attachment: Attachment | undefined = this.getAttachment(appeal, fileId);
+
+        if (!attachment) {
+            return false;
+        }
+
+        // User must be creating a new appeal.
+        if (!appeal.createdBy) {
+            return true;
+        }
+
+        // Appeal was loaded from API
+        if (appeal.createdBy?._id !== userProfile.id) {
+            return false;
+        }
+
+        return true;
+    }
+
+    private renderForbiddenError(res: Response): void {
+        res.status(FORBIDDEN);
+        res.render(customErrorTemplate, {
+            heading: 'You are not authorised to download this document'
+        });
+    }
+
+    private getAttachment(appeal: Appeal, fileId: string): Attachment | undefined {
+
+        if (!fileId) {
+            throw Error('File id must not be null');
+        }
+
+        const attachments: Attachment[] | undefined = appeal.reasons.other.attachments;
+        return attachments && attachments.find(attachement => attachement.id === fileId);
```

Comment:
minor spelling mistake, should be `attachment` - to be honest, i find myself doing the same :D

Reply:
🤦